File: Exceptions/ReallyBadCode/CreateExceptions.vb
Public Module CreateExceptions
	Public Sub ThrowIt(ByVal switch As String)
		If switch = "2" Then
			Throw New ArgumentException("Nothing works.", "switch")
		End If
	End Sub

	Public Sub ThrowNonpublicException()
		Throw New NonPublicCustomException()
	End Sub

	Public Sub CreateArgumentExceptionWithWrongArgumentName( _
		ByVal index As Integer)
		If index = 0 Then
			Throw New ArgumentException("index", "This is a bad index value.")
		End If
	End Sub

	Public Sub CreateArgumentExceptionWithNoParameters(ByVal index As Integer)
		If index = 0 Then
			Throw New ArgumentException()
		End If
	End Sub
End Module

File: Exceptions/CodeToImprove/CreateExceptions.cs
using System;

namespace CodeToImprove
{
	public static class CreateExceptions
	{
		public static void ThrowIt(string value)
		{
			if (value == "2")
			{
				throw new ArgumentException("Nothing works", nameof(value));
			}
		}

		public static void ThrowNonPublicException() => throw new NonPublicCustomException();

		public static void CreateArgumentExceptionWithWrongArgumentName(int index)
		{
			if(index == 0)
			{
				throw new ArgumentException("index", "This is a bad index value.");
			}
		}

		public static void CreateArgumentExceptionWithNoParameters(int index)
		{
			if (index == 0)
			{
				throw new ArgumentException();
			}
		}
	}
}